strawberry/union.py

```diff
@@ -135,9 +135,16 @@ def _resolve_union_type(
 
             from strawberry.types.types import TypeDefinition
 
-            # Make sure that the type that's passed in is an Object type
+            # If the type given is not an Object type, try resolving using `is_type_of`
+            # defined on the union's inner types
             if not hasattr(root, "_type_definition"):
-                # TODO: If root=python dict, this won't work
+                for inner_type in type_.types:
+                    if inner_type.is_type_of is None:
+                        break
+                    elif inner_type.is_type_of(root, info):
+                        return inner_type.name
```

Comment:
could we simplify this to:
```suggestion
                    if inner_type.is_type_of is not None and inner_type.is_type_of(root, info):
                        return inner_type.name
```
?